### DISCUSSION TITLE
How to perform parallel run in multiple nodes/machine in Jenkins

### DISCUSSION THREAD

**Discussion by bilarallen** [cypress-io/cypress#26117] [Questions and Help] [answered]
Hi! I hope someone can give an idea how we can perform parallel execution in multiple node/machine in Jenkins pipeline. I followed the instruction here https://github.com/cypress-io/cypress-example-kitchensink/blob/master/Jenkinsfile for setting up Jenkins parallel execution. But it seems that it only execute in one machine.
![image](https://user-images.githubusercontent.com/65072341/225206308-bc9c0e44-2266-4647-918a-29756f25928c.png)
![image](https://user-images.githubusercontent.com/65072341/225206337-656fbad5-587c-468a-9914-77cad5b0b406.png)

Whenever I check the logs I see that every stages are being run in parallel but only in one machine
![image](https://user-images.githubusercontent.com/65072341/225206378-1dc226bd-3d05-485d-b171-da330fe19045.png)

My JenkinsPipeline is declarative and here is the setup

```
pipeline {
    agent {
        docker {
            label 'default'
            image 'cypress/base:14.16.0'
        
        }
    }
    stages {
        stage('Install Dependencies') { 
            steps {
                echo "Running build ${env.BUILD_ID} on ${env.JENKINS_URL}"
                sh 'npm ci'
                sh 'npm run cy:verify'
            }
        }
        
        stage('CBS UI Test') { 
            environment {
                CYPRESS_RECORD_KEY = credentials('cypress-record-key')
                CYPRESS_trashAssetsBeforeRuns = 'false'
            }

            parallel {
                stage('Tester A') {
                    steps {
                        echo "Running build ${env.BUILD_ID} on Tester A"
                        sh "NO_COLOR=1 xvfb-run -a npx cypress-cloud run --parallel --record --key ${env.CYPRESS_RECORD_KEY}"
                    }
                }

                stage('Tester B') {
                    steps {
                        echo "Running build ${env.BUILD_ID} on Tester B"
                       sh "NO_COLOR=1 xvfb-run -a npx cypress-cloud run --parallel --record --key ${env.CYPRESS_RECORD_KEY}"
                    }
                }

                stage('Tester C') {
                    steps {
                        echo "Running build ${env.BUILD_ID} on Tester C"
                       sh "NO_COLOR=1 xvfb-run -a npx cypress-cloud run --parallel --record --key ${env.CYPRESS_RECORD_KEY}"
                    }
                }

            }
           
        }
    }
}
```
Our built-in node has 6 executors and this is only the node where we can run our test as we specify by its label
![image](https://user-images.githubusercontent.com/65072341/225206756-00fc3529-3b23-4309-9acc-5f43f2f97f3a.png)

Our goal is to run it in parallel either in multiple nodes (where we will set up different machines) or in multiple executors using the same node.

Please take note that our Jenkins is dockerized that's why in my Jenkins pipeline I need to use the docker cypress base image.

One thing to add is that even though our test is in parallel right now running our test in locally is bit faster compared when running it in CI using Cypress parallelization
![image](https://user-images.githubusercontent.com/65072341/225206801-ad5f15cc-5312-4ce5-bc18-d11f7eafede5.png)
Using Cypress parallel execution it took 16 mins. to finish the whole test.

Hoping someone can help or give us a code snippet to follow based on our setup. Thanks!

**Comment by 3mrfouad** [comment]
Hello @bilarallen did you get a breakthrough here? I have same issue trying to get `Jenkins` running multiple agents with `Cypress Cloud`.

**Reply by bilarallen** [reply, marked as answer]
Hey @3mrfouad yep I was able to resolve this issue. Basically, you need different nodes/agents to achieve parallel run in different machines. We are using AWS ec2 instance as VMS a.k.a `ec2-fleet` then you need to specify the agent label. Here's the code snippet I have for my jenkins file:

```
pipeline {
    agent none
    stages {
        stage('CBS UI Test') { 
            environment {
                CYPRESS_RECORD_KEY = credentials('cypress-record-key')
                CYPRESS_trashAssetsBeforeRuns = 'false'
            }

            parallel {
                stage('Tester A') {
                   agent {
                        docker {
                        label 'ec2-fleet-cypress'
                        image 'cypress/base:14.16.0'
                        }
                    }
                    steps {
                        echo "Running build ${env.BUILD_ID} on Tester A"
                        sh "NO_COLOR=1 xvfb-run -a npx cypress-cloud run --parallel --record --key ${env.CYPRESS_RECORD_KEY}"
                    }
                }

                stage('Tester B') {
                    agent {
                        docker {
                        label 'ec2-fleet-cypress'
                        image 'cypress/base:14.16.0'
                        }
                    }
                    steps {
                        echo "Running build ${env.BUILD_ID} on Tester B"
                       sh "NO_COLOR=1 xvfb-run -a npx cypress-cloud run --parallel --record --key ${env.CYPRESS_RECORD_KEY}"
                    }
                }

                stage('Tester C') {
                    agent {
                        docker {
                        label 'ec2-fleet-cypress'
                        image 'cypress/base:14.16.0'
                        }
                    }
                    steps {
                        echo "Running build ${env.BUILD_ID} on Tester C"
                       sh "NO_COLOR=1 xvfb-run -a npx cypress-cloud run --parallel --record --key ${env.CYPRESS_RECORD_KEY}"
                    }
                }

            }
           }
        }
    }
}
```
